### PR TITLE
TST: Check map function works with StringDtype (#40823)

### LINKED ISSUE #40823: BUG: Pandas map function does not work properly with StringDtype
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd


df1 = pd.DataFrame(
    {
        "col1": [pd.NA, "foo", "bar"]
    }, index=["id1", "id2", "id3"], dtype=pd.StringDtype()
)

df2 = pd.DataFrame(
    {
        "id": ["id4", "id2", "id1"]
    }, dtype=pd.StringDtype()
)

df2["col1"] = df2["id"].map(df1["col1"])

print(df2.head())

```

#### Problem description

Traceback:
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\Program Files\JetBrains\PyCharm 2020.3.2\plugins\python\helpers\pydev\_pydev_bundle\pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm 2020.3.2\plugins\python\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:/Users/XXX/AppData/Roaming/JetBrains/PyCharm2020.3/scratches/scratch_22.py", line 16, in <module>
    df2["col1"] = df2["id"].map(df1["col1"])
  File "C:\UsersXXX\AppData\Local\pypoetry\Cache\virtualenvs\basel-w19Rd76b-py3.9\lib\site-packages\pandas\core\series.py", line 3909, in map
    new_values = super()._map_values(arg, na_action=na_action)
  File "C:\Users\XXX\AppData\Local\pypoetry\Cache\virtualenvs\basel-w19Rd76b-py3.9\lib\site-packages\pandas\core\base.py", line 908, in _map_values
    new_values = algorithms.take_1d(mapper._values, indexer)
  File "C:\Users\XXX\AppData\Local\pypoetry\Cache\virtualenvs\basel-w19Rd76b-py3.9\lib\site-packages\pandas\core\algorithms.py", line 1699, in take_nd
    return arr.take(indexer, fill_value=fill_value, allow_fill=allow_fill)
  File "C:\Users\XXX\AppData\Local\pypoetry\Cache\virtualenvs\basel-w19Rd76b-py3.9\lib\site-packages\pandas\core\arrays\_mixins.py", line 78, in take
    return self._from_backing_data(new_data)
  File "C:\Users\XXX\AppData\Local\pypoetry\Cache\virtualenvs\basel-w19Rd76b-py3.9\lib\site-packages\pandas\core\arrays\numpy_.py", line 190, in _from_backing_data
    return type(self)(arr)
  File "C:\Users\XXX\AppData\Local\pypoetry\Cache\virtualenvs\basel-w19Rd76b-py3.9\lib\site-packages\pandas\core\arrays\string_.py", line 195, in __init__
    self._validate()
  File "C:\Users\XXX\AppData\Local\pypoetry\Cache\virtualenvs\basel-w19Rd76b-py3.9\lib\site-packages\pandas\core\arrays\string_.py", line 200, in _validate
    raise ValueError("StringArray requires a sequence of strings or pandas.NA")
ValueError: StringArray requires a sequence of strings or pandas.NA

This issue can be circumvented if I do:
```python
df2["col1"] = df2["id"].map(df1["col1"].astype("object"))
print(df2.head())
```
Gives:    
    id  col1
0  id4   NaN
1  id2   foo
2  id1  \<NA\>

Which is a combination of np.nan and pd.NA

If I do then a conversion back to pd.StringDtype:
```python
df2["col1"] = df2["id"].map(df1["col1"].astype("object")).astype(pd.StringDtype())
print(df2.head())
```
I get the expected output: 
    id  col1
0  id4  \<NA\>
1  id2   foo
2  id1  \<NA\>

But this is ugly as hell.

#### Expected Output
    id  col1
0  id4   \<NA\>
1  id2   foo
2  id1  \<NA\>
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f2c8480af2f25efdbd803218b9d87980f416563e
python           : 3.9.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 79 Stepping 1, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252
pandas           : 1.2.3
numpy            : 1.20.2
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 52.0.0
Cython           : None
pytest           : 6.2.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.1
sqlalchemy       : 1.4.3
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by mdhsieh**
Checks map function output on 2 DataFrames with data type StringDtype.

- [ ] closes #40823
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/frame/methods/test_map.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import pandas as pd
+import pandas._testing as tm
+
+
+class TestMap:
```
Comment (jreback) [inline pandas/tests/frame/methods/test_map.py:5]: we have extensive map tests here: pandas/tests/apply/test_series_apply.py

see if you can find something similar. you can use the fixture `any_string_dtype` to parmeterize this over various kinds of strings.

File: pandas/tests/frame/methods/test_map.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+        df2["col1"] = df2["id"].map(df1["col1"])
+
+        result = df2
+        expected = pd.DataFrame(
+            {"id": ["id4", "id2", "id1"], "col1": [pd.NA, "foo", pd.NA]},
+            dtype=pd.StringDtype(),
+        )
+
+        tm.assert_frame_equal(result, expected)
```
Comment (jreback) [inline pandas/tests/frame/methods/test_map.py:24]: this is a series method, so test on that.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_map.py (hunk 1 above)
Comment (mdhsieh) [reply pandas/tests/frame/methods/test_map.py]: Ok moved. Should the fixture be used like this? 
```
# parameter
test_series_map_stringdtype(any_string_dtype):
    ser1 = Series(
        # ...
        # then set here for all Series
        dtype=any_string_dtype,
    )
 ```

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mdhsieh

**Pull request merged**